agent/consul/state/intention.go

```diff
@@ -936,3 +930,91 @@ func intentionMatchGetParams(entry structs.IntentionMatchEntry) ([][]interface{}
 	result = append(result, []interface{}{entry.Namespace, entry.Name})
 	return result, nil
 }
+
+// IntentionTopology returns the upstreams or downstreams of a service. Upstreams and downstreams are inferred from
+// intentions. If intentions allow a connection from the target to some candidate service, the candidate service is considered
+// an upstream of the target.
+func (s *Store) IntentionTopology(ws memdb.WatchSet,
+	target structs.ServiceName, downstreams bool, defaultDecision acl.EnforcementDecision) (uint64, structs.ServiceList, error) {
+	tx := s.db.ReadTxn()
+	defer tx.Abort()
+
+	var maxIdx uint64
+
+	// If querying the upstreams for a service, we first query intentions that apply to the target service as a source.
+	// That way we can check whether intentions from the source allow connections to upstream candidates.
+	matchType := structs.IntentionMatchSource
+	if downstreams {
+		matchType = structs.IntentionMatchDestination
+	}
+	entry := structs.IntentionMatchEntry{
+		Namespace: target.NamespaceOrDefault(),
+		Name:      target.Name,
+	}
+	index, intentions, err := compatIntentionMatchOneTxn(tx, ws, entry, matchType)
+	if err != nil {
+		return 0, nil, fmt.Errorf("failed to query intentions for %s", target.String())
+	}
+	if index > maxIdx {
+		maxIdx = index
+	}
+
+	// Reset the matchType since next it is used for evaluating the upstreams or downstreams against a set of intentions.
+	// When evaluating upstreams, the match type is now destination because we are evaluating upstream candidates
+	// as eligible destinations for intentions that have the target service as a source.
+	// The reverse is true for downstreams.
+	matchType = structs.IntentionMatchDestination
```

Comment:
I wonder if this would be clearer if up front you just declared two vars like `matchTypeUp` and `matchTypeDown` and only did the `if downstreams { ... } else { ... }` conditional once rather than mutating a variable midway through.

Reply:
I didn't fully understand what you were recommending. I ended up renaming these so that the matchType is prefixed with the intended use.